services/alerts/src/AlertsManagerContext.ts

```diff
@@ -0,0 +1,62 @@
+import React from 'react'
+import { Alert, AlertsManager, AlertsManagerAlert } from './types'
+
+type SetAlertsFunction = React.Dispatch<React.SetStateAction<Alert[]>>
+
+export const makeAlertManager = (
+    setAlerts: SetAlertsFunction
+): AlertsManager => {
+    let id = 0
+
+    const remove = (alert: Alert) => {
+        setAlerts(alerts => {
+            const idx = alerts.findIndex(a => a === alert)
+
+            if (idx === -1) {
+                return alerts
+            }
+
+            return [
+                ...alerts.slice(0, idx),
+                ...alerts.slice(idx + 1, alerts.length),
+            ]
+        })
+    }
+    const add = (alert: Alert) => {
+        setAlerts(alerts => {
+            if (alerts.some(a => a === alert)) {
+                return alerts
+            }
+
+            id++
+
+            const alertManagerAlert: AlertsManagerAlert = {
+                ...alert,
+                id,
+                get remove() {
+                    return () => {
+                        alert.options.onHidden && alert.options.onHidden()
+                        remove(this)
```

Comment:
This getter confused me, though I think I see why it was necessary.  I believe it could simply refer to the `alertManagerAlert` variable (untested):

```js
const alertManagerAlert: AlertManagerAlert = {
  ...alert,
  id,
  remove: () => {
    alert.options.onHidden && alert.options.onHidden()
    remove(alertManagerAlert)
  }
}
```

If not (and maybe even if so) I think we could rewrite `remove` to accept an id and just call `remove(id)` instead of relying on object integrity, now that we have a unique id available.

Reply:
It confused me as well. Especially with the proposal to use the id as the argument to remove alerts, I think this is unnecessary

Reply:
Agreed, if we use the id to identify an alert instead of checking for reference equality, we can and should make this simpler.